### PR TITLE
RTL8723BS: add support for OBDA8723 - Bluetooth

### PR DESCRIPTION
**Pull request by sorhtyre**
Compiled and tested against 4.9.13 and 4.10.11 - 4.10.14 kernels. No issued reported.

### CODE REVIEW

**Comment by vpeter4** [comment]
I think service file should be part of RTL8723BS_BT package?

**Comment by sorhtyre** [comment]
I considered adding it, but was uncertain if it was inclusive enough as I have only tested it against the LibreELEC distribution. I can easily add it if they is the recomendation.

File: projects/Generic/filesystem/usr/lib/systemd/system/rtk_bt.service [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Realtek RTL8723BS bluetooth service
+ConditionPathExists=/dev/ttyS1
```
Comment (MilhouseVH) [inline projects/Generic/filesystem/usr/lib/systemd/system/rtk_bt.service]: `/dev/ttyS1` is going to exist on pretty much every PC, regardless of whether it has this chipset or not, and as such it isn't a suitable condition on which to base the loading of this driver.

Comment (sorhtyre) [inline projects/Generic/filesystem/usr/lib/systemd/system/rtk_bt.service]: Understood. I will research a better method that is more selective.

File: packages/linux-drivers/RTL8723BS_BT/package.mk [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+       ARCH=$TARGET_KERNEL_ARCH \
+       KSRC=$(kernel_path) \
+       CROSS_COMPILE=$TARGET_PREFIX \
+       CONFIG_POWER_SAVING=n
+}
+
+makeinstall_target() {
+  mkdir -p $INSTALL/usr/lib/firmware/rtl_bt
+    cp -p firmware/rtlbt_config $INSTALL/usr/lib/firmware/rtl_bt
+    cp -p firmware/rtlbt_fw $INSTALL/usr/lib/firmware/rtl_bt
```
Comment (MilhouseVH) [inline packages/linux-drivers/RTL8723BS_BT/package.mk]: Does the firmware have to have such a generic name? Wouldn't it be better to include the name of the respective chipset, as is common with other in-tree RTL firmware, eg.:
```
-rw-rw-r--  1 neil neil  13904 Apr 12 05:12 rtl8188eufw.bin
-rw-rw-r--  1 neil neil  16192 Apr 12 05:12 rtl8192cfw.bin
-rw-rw-r--  1 neil neil  16332 Apr 12 05:12 rtl8192cfwU_B.bin
-rw-rw-r--  1 neil neil  14818 Apr 12 05:12 rtl8192cfwU.bin
-rw-rw-r--  1 neil neil  16116 Apr 12 05:12 rtl8192cufw_A.bin
```
etc.? (the above is from the [linux-firmware](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/rtlwifi) tree)

Comment (sorhtyre) [inline packages/linux-drivers/RTL8723BS_BT/package.mk]: Agreed. I will verfiy tonight.

Comment (stefansaraev) [inline packages/linux-drivers/RTL8723BS_BT/package.mk]: yet another out-of tree rtl self firmware loading nonse. and @MilhouseVH no this is supposed to be as it is already done.

ref:
https://github.com/sorhtyre/rtl8723bs_bt/blob/BOXSTCK1A8LFC/Makefile#L17-L18
https://github.com/sorhtyre/rtl8723bs_bt/blob/BOXSTCK1A8LFC/hciattach_rtk.c#L52-L53
https://github.com/sorhtyre/rtl8723bs_bt/blob/BOXSTCK1A8LFC/hciattach_rtk.c#L1422
https://github.com/sorhtyre/rtl8723bs_bt/blob/BOXSTCK1A8LFC/hciattach_rtk.c#L1507

so, no, as it's not using kernel's generic fw loader.

File: projects/Generic/filesystem/usr/lib/systemd/system/rtk_bt.service (hunk 1 above)
Comment (sorhtyre) [inline projects/Generic/filesystem/usr/lib/systemd/system/rtk_bt.service]: I believe I can replace the condition check with:
ConditionPathExists=/sys/devices/platform/80860F0A:00/OBDA8723:00

File: packages/linux-drivers/RTL8723BS_BT/package.mk (hunk 2 above)
Comment (sorhtyre) [inline packages/linux-drivers/RTL8723BS_BT/package.mk]: I will rename the firmware, but I want to ensure it's not confused with firmware included in the rtl8723bs* firmware from the kernel-firmware package, which oddly enough does not work.

File: projects/Generic/filesystem/usr/lib/systemd/system/rtk_bt.service (hunk 1 above)
Comment (MilhouseVH) [inline projects/Generic/filesystem/usr/lib/systemd/system/rtk_bt.service]: That looks better. :)

File: packages/linux-drivers/RTL8723BS_BT/package.mk (hunk 2 above)
Comment (MilhouseVH) [inline packages/linux-drivers/RTL8723BS_BT/package.mk]: Yes, supporting Realtek with these out of tree drivers and firmware is a PITA. It's a shame the in-tree kernel firmware can't be used (it's actually installed from [wlan-firmware](https://github.com/LibreELEC/wlan-firmware/tree/master/firmware/rtlwifi), but that could change as there's so much overlap with the linux-firmware repo). Giving the firmware in this PR a somewhat unique/recognisable name is just a nicety, and might be useful in future if anyone else wants to support additional RTL hardware and needs to include their own firmware (though hopefully not... 😄 )

**Comment by sorhtyre** [comment]
All updates are complete and posted. I will rebuild everything to ensure I didn't break it.

**Comment by MilhouseVH** [comment]
Looks fine, I think... although obviously only you can test. :)

I'll include this PR in tonight's LE9 test builds [here](http://forum.kodi.tv/showthread.php?tid=298462) - please can you give it a go later this evening?

Assuming this works with LE8 (and I think it should) then maybe a backport would be good too (ping @chewitt), although it is unfortunate this is based on yet more out-of-tree Realtek drivers/firmware, as they're often a pain when moving to a major new kernel, so I'll leave the final merge decision to @lrusak & @chewitt.

File: packages/linux-drivers/RTL8723BS_BT/package.mk [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+PKG_NAME="RTL8723BS_BT"
+PKG_VERSION="BOXSTCK1A8LFC"
```
Comment (MilhouseVH) [inline packages/linux-drivers/RTL8723BS_BT/package.mk]: One thing... can you change the `PKG_VERSION` to a specific rev, such as `6a6f1da`.

Using a branch label as version means everyone will be building with different source code (depending on when they download the tarball, and won't receive any updates) should you check in any additional commits.

**Comment by chewitt** [comment]
RTL8723BS has mainline support coming in Linux 4.12 so I'd prefer to see a curated patch that backports those commits to 4.9/4.11 instead of using an 'out of tree' version. I'd guess (or hope) that firmware in the kernel firmware repo works with the kernel driver?

see: https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/log/?id=refs%2Ftags%2Fnext-20170505&qt=grep&q=RTL8723BS

**Comment by MilhouseVH** [comment]
I have included this PR in a test build [#0506](http://forum.kodi.tv/showthread.php?tid=298462&pid=2582526#pid2582526) but would support @chewitts comment that if a kernel 4.12 backport is possible that would be preferable to a separate package etc. I'll drop this PR from future builds until the way ahead is decided.

**Comment by sorhtyre** [comment]
@MilhouseVH, I have completed all tests. I needed to add a systemd timer to load the service 10 seconds after boot. I also updated the package version as recommended.

@chewitt, I will look into adding this to Linux 4.12...

**Comment by MilhouseVH** [comment]
Pretty sure the 10-second delay timer is a really bad idea - if you need to delay loading of the driver until some hardware or software component is available then add that as a dependency, not a random time interval which may or may not be sufficient depending on how quickly (or slowly) the system boots.

**Comment by sorhtyre** [comment]
I need to wait for sysfs to populate... I will figure out a better method.

If I didn't have so much homework, I could have completed this by now. By which I mean, I apologize for dragging this out...

File: packages/linux-drivers/RTL8723BS_BT/package.mk [hunk 4]
```diff
@@ -0,0 +1,59 @@
+################################################################################
+#      This file is part of LibreELEC - https://libreelec.tv
+#      Copyright (C) 2016-2017 Team LibreELEC
```
Comment (chewitt) [inline packages/linux-drivers/RTL8723BS_BT/package.mk]: technically this would be 2017 only as it's a new file

Comment (sorhtyre) [inline packages/linux-drivers/RTL8723BS_BT/package.mk]: Corrected.

I have tested the firmware included in the Linux kernel (rtl8723bs_bt.bin) and found that is does not work. 
https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/rtlwifi

**Comment by chewitt** [comment]
@sorhtyre I have a strong preference for backporting vs. adding an out of tree package. Please test the 4.12 firmware with the 4.12 driver (not 4.12 firmware and out of tree driver). If that test concludes something is broken it needs to be reported/investigated upstream so that things are fixed before 4.12 is released.

**Comment by MilhouseVH** [comment]
@sorhtyre can you test this 4.12-rc1 based build that includes RTL8273BS as an in-tree module?

http://forum.kodi.tv/showthread.php?tid=298462&pid=2588196#pid2588196

It's probably lacking any required firmware, as we're not installing the firmware from [linux-firmware/rtlwifi](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/rtlwifi) but that can be added easily enough if you can confirm which firmware is required etc.

**Comment by sorhtyre** [comment]
@MilhouseVH, will do. I'm studying for a midterm this weekend, so it might not be until tomorrow.

I did, however, perform tests using the firmware before 4.12-rc1 dropped and none of the firmware worked for the bluetooth interface. I find it odd that one of the files is named rtl8723bs_bt.bin. Which suggests it is firmware for the device. I will send out inquiries to find out why it was included.

I believe that I have also made an error in including this package in the linux-driver section, as it really should be in linux-firmware folder. How should I proceed?

The package as-is does not seem to requre kernel headers and is independent of the running kernel. The only part that needs to be upstreamed is the rfkill-gpio patch and modifications to the config file to include required modules.

The next part is to request an upstream of the rtk_hciattach.c to the BlueZ package.

**Comment by MilhouseVH** [comment]
I prefer this in the linux-driver section. There are some packages (such as brcmfmac*) that maybe shouldn't be in linux-firmware but should be in linux-drivers. The idea of the linux-firmware category (IMHO) is that it is for packages that install *only* blobs, and is not for packages that install drivers, services, scripts (and maybe a blob or two).

**Comment by chewitt** [comment]
Agreed. Firmware is firmware only. Anything else is considered to be a driver package.

**Comment by sorhtyre** [comment]
Understood. I will maintain the package in linux-driver.

I have tested the 4.12-rc1 based build you linked and found that I received a kernel panic upon boot. I believe this is due to not loading the rtl8723bs_nic.bin.

kernel panic

**Comment by MilhouseVH** [comment]
As a temporary measure can you install [rtl8723bs_bs.bin](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/rtlwifi/rtl8723bs_bt.bin) in `/storage/.config/firmware` (or possibly, `/storage/.config/firmware/rtlwifi`) and see if that helps. I'll include this firmware in the next 4.12-rc2 build later this week.

**Comment by MilhouseVH** [comment]
The 4.12-rc2 [test build](http://forum.kodi.tv/showthread.php?tid=298462&pid=2590172#pid2590172) now includes all rtl8723bs* firmware (that exists upstream)

**Comment by sorhtyre** [comment]
@MilhouseVH, the 4.12-rc2 test build works! Do you still want the `journalctl -a | pastebinit` output?

**Comment by chewitt** [comment]
Yes please.

**Comment by sorhtyre** [comment]
http://sprunge.us/egUW

**Comment by chewitt** [comment]
@sorhtyre thanks for the original proposal and testing effort. I'm going to close this PR down as it's superseded with the inclusion of RTL8723BS in Linux 4.12 via #1629.

**Pull request closed without merging**